api/v1alpha1/azurevalidator_types.go

```diff
@@ -21,20 +21,61 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
 // AzureValidatorSpec defines the desired state of AzureValidator
 type AzureValidatorSpec struct {
-	Auth          AzureAuth `json:"auth"`
-	DefaultRegion string    `json:"defaultRegion"`
+	// Rules for validating role assignments in Azure RBAC.
+	RoleAssignmentRules []RoleAssignmentRule `json:"roleAssignmentRules"`
+}
+
+// RoleAssignmentRule is a rule that validates that a desired role assignment
+// exists within a subscription. The role is specified as its role name (e.g.
+// "Contributor"). The validator takes care of looking up the details of the
+// role definition associated with the role name automatically.
+type RoleAssignmentRule struct {
+	Role               Role   `json:"role"`
+	ServicePrincipalID string `json:"servicePrincipalId"`
+	SubscriptionID     string `json:"subscriptionId"`
+}
+
+// GetRoleName returns the name of the role in the role assignment. Returns nil if the user did not
+// specify it in the spec.
+func (r RoleAssignmentRule) GetRoleName() *string {
```

Comment:
None of these "getters" should be required. You can always access a `RoleAssignmentRule`'s fields directly.

Reply:
I chose to use an interface and getters because this is the pattern used in AWS. We need the interface for testability. You can see in the AWS plugin codebase a spot where we allow anything satisfying the rule interface to be used as an argument:

```go
func (s *IAMRuleService) ReconcileIAMRoleRule(rule iamRule) (*types.ValidationResult, error) {
```

And in its tests, a spot where we use a mock as an argument:

```go
type testCase struct {
	name           string
	rule           iamRule
	expectedResult types.ValidationResult
	expectedError  error
}

...

result, err := iamService.ReconcileIAMRoleRule(c.rule)
```

However, I'm glad you pointed this out to me because it made me realize that I only need to use interfaces at the top level. I originally tried making a `role` interface for a `GetRole` getter to return, because I thought I needed to use interfaces all the way down. This pattern didn't work. 

Now I realize I don't need to make a `role` interface and have a method signature use that interface as its return type. I can just have my methods return parts of the spec. It'll all still be testable. That's what the AWS plugin codebase does, and I'll need to do it that way too to support the "able to validate more than one rule per validator" requirement.

Reply:
I addressed this by adding a `GetRole` getter back in, but making it return the actual struct from my spec (`Role`). I should be able to continue now as I make it support multiple roles within a rule, and then multiple rules within a validator.